### PR TITLE
Improves performance of show command 

### PR DESCRIPTION
**Pull request by cgranleese-r7**
This PR addresses the slow load times on the `show` command within msfconsole.

The changes I have added will now allow the `show` command to make use of the metasploit cache. 

Previously the `show` command was slow when returning bigger lists of results e.g. `show exploits` was taking ~13 seconds, whereas now it's under a second.

This change was applied to the following commands:
- `show encoders`
- `show nops`
- `show exploits`
- `show payloads`
- `show auxiliary`
- `show post`
- `show evasion`
- `show favorites`

## Performance stats
Time in seconds for each command to finish, using the new `time` command:
| Command |  Old method  |  New method  | 
|:-:|:-:|:-:|
|  `show encoders` with active module |  0.03 |  0.06 |   
|  `show payloads` with active module |  2.7 |  2.9 |
| `show encoders` |  0.02 |  0.04 |  
|  `show payloads` |  1.0 |  0.3 |   
|  `show exploits` |  13.8 |  0.9 |
|  `show auxiliary` |  3.2 |  0.4 |
|  `show post` |  1.9 |  0.14 |
|  `show evasion` |  0.05 |  0.03 |
|  `show favorites` |  0.02 |  0.03 |

## Before
show_exploit_performance_master
## After 
show_exploit_performance

### Side note
My IDE kept removing two lines where there was a line with indentation but no code, couldn't seem to get it to not edit those lines ¯\\\_(ツ)_/¯

## Verification

- [ ] Start `msfconsole`
- [ ] ` use exploit/windows/smb/ms17_010_eternalblue`
- [ ] Go through the list of commands above and test performance and functionality
- [ ] **Verify** the output content is the same as what was being returned on `master`
- [ ] **Verify** the performance of the commands output is now faster.

### CODE REVIEW

**Comment by adfoster-r7** [comment]
Before
```
[+] Command "show exploits" completed in 14.455557000008412 seconds
```

After
```
[+] Command "show exploits" completed in 0.4832089999690652 seconds
```

File: lib/msf/ui/console/command_dispatcher/modules.rb [hunk 1]
```diff
@@ -1527,57 +1528,66 @@ def show_plugins # :nodoc:
...
-            count = -1
+          # @param [table_name] used to name table
+          # @param [module_filter] this will either be a modules fullname, or it will be an Array(show payloads/encoders)
+          # or a Hash(show favorites) containing fullname
+          # @param [compatible_mod] handles logic for if there is an active module when the
+          # `show` command is run
+          #
+          # Handles the filtering of modules that will be generated into a table
+          def show_module_metadata(table_name, module_filter)
+            count = 0
```
Comment (adfoster-r7) [inline lib/msf/ui/console/command_dispatcher/modules.rb:1539]:
```suggestion
            count = -1
```

Metasploit's tables start from 0.

Master:
```
msf6 > show encoders

Encoders
========

   #   Name                          Disclosure Date  Rank       Check  Description
   -   ----                          ---------------  ----       -----  -----------
   0   cmd/brace                                      low        No     Bash Brace Expansion Command Encoder
```

This branch:
```
msf6 > show encoders

Encoders
========

   #   Name                                  Disclosure Date  Rank       Check  Description
   -   ----                                  ---------------  ----       -----  -----------
   1   encoder/cmd/brace                                      low        No     Bash Brace Expansion Command Encoder
```

Comment (cgranleese-r7) [reply lib/msf/ui/console/command_dispatcher/modules.rb:1539]: I didn't realise that and thought this was just an error, makes sense now as to why it was beginning at `-1` 👍 

Will get that added back in.

Comment (adfoster-r7) [reply lib/msf/ui/console/command_dispatcher/modules.rb]: Looked into this a bit more, sessions start at 1, but these tables start at 0 after this PR was introduced
https://github.com/rapid7/metasploit-framework/pull/11819

**Comment by adfoster-r7** [comment]
## Release Notes

Improves the performance of the various `show` commands within the console. For instance `show exploits` now takes ~0.5 seconds instead of ~14 seconds

**Pull request merged**